cou/apps/base.py

```diff
@@ -132,42 +132,44 @@ def __str__(self) -> str:
         return yaml.dump(summary, sort_keys=False)
 
     @property
-    def apt_source_codename(self) -> Optional[OpenStackRelease]:
+    def apt_source_codename(self) -> OpenStackRelease:
         """Identify the OpenStack release set on "openstack-origin" or "source" config.
 
-        :raises ApplicationError: If os_origin_parsed is not a valid OpenStack release or os_origin
-            is in an unexpected format (ppa, url, etc).
-        :return: OpenStackRelease object or None if the app doesn't have os_origin config.
-        :rtype: Optional[OpenStackRelease]
+        :raises ApplicationError: When origin setting is not valid.
+        :return: OpenStackRelease object.
+        :rtype: OpenStackRelease
         """
-        os_origin_parsed: Optional[str]
-        # that means that the charm doesn't have "source" or "openstack-origin" config.
-        if self.origin_setting is None:
-            return None
+        #  means that the charm doesn't have origin setting config or is using empty string.
+        if not self.os_origin:
+            return self.current_os_release
 
-        # Ex: "cloud:focal-ussuri" will result in "ussuri"
         if self.os_origin.startswith("cloud"):
-            *_, os_origin_parsed = self.os_origin.rsplit("-", maxsplit=1)
-            try:
-                return OpenStackRelease(os_origin_parsed)
-            except ValueError as exc:
-                raise ApplicationError(
-                    f"'{self.name}' has an invalid '{self.origin_setting}': {self.os_origin}"
-                ) from exc
-
-        elif self.os_origin == "distro":
+            return self._extract_from_uca_source()
+
+        if self.os_origin == "distro":
             # find the OpenStack release based on ubuntu series
-            os_origin_parsed = DISTRO_TO_OPENSTACK_MAPPING[self.series]
-            return OpenStackRelease(os_origin_parsed)
+            return OpenStackRelease(DISTRO_TO_OPENSTACK_MAPPING[self.series])
 
-        elif self.os_origin == "":
-            return None
+        # probably because user set a ppa or an url
+        raise ApplicationError(
+            f"'{self.name}' has an invalid '{self.origin_setting}': {self.os_origin}"
+        )
 
-        else:
-            # probably because user set a ppa or an url
+    def _extract_from_uca_source(self) -> OpenStackRelease:
+        """Extract the OpenStack release from Ubuntu Cloud Archive (UCA) sources.
+
+        :raises ApplicationError: When origin setting is not valid.
+        :return: OpenStackRelease object
+        :rtype: OpenStackRelease
+        """
+        # Ex: "cloud:focal-ussuri" will result in "ussuri"
+        *_, os_origin_parsed = self.os_origin.rsplit("-", maxsplit=1)
+        try:
+            return OpenStackRelease(os_origin_parsed)
```

Comment:
If `openstack-origin` value should always follow the same format as `"cloud:focal-ussuri"`, we can move the parsing into the try block, plus removing `*`, to catch value error early (`os_origin` doesn't contain `-`)
```suggestion
        try:
            # Ex: "cloud:focal-ussuri" will result in "ussuri"
            _, os_origin_parsed = self.os_origin.rsplit("-", maxsplit=1)
            return OpenStackRelease(os_origin_parsed)
```